Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,413 @@
+                ValidatePathCache(confinerStateChanged, frustumHeight);
+
+                var extra = GetExtraState<VcamExtraState>(vcam);
+                Vector3 displacement = ConfinePoint(state.CorrectedPosition);
+                if (VirtualCamera.PreviousStateIsValid && deltaTime >= 0)
+                { 
+                    float displacementAngle = Vector2.Angle(extra.m_previousDisplacement, displacement);
+                    if (m_CornerDampingIsOn || m_Damping > 0 && displacementAngle > m_CornerAngleTreshold)
+                    {
+                        Vector3 delta = displacement - extra.m_previousDisplacement;
+                        var deltaDamped = 
+                            Damper.Damp(delta, m_Damping / m_CornerDampingSpeedup, deltaTime);
+                        displacement = extra.m_previousDisplacement + deltaDamped;
+
+                        m_CornerDampingSpeedup = displacementAngle < 1f ? 2f : 1f;
+                        m_CornerDampingIsOn = displacementAngle > UnityVectorExtensions.Epsilon ||
+                                              delta.sqrMagnitude > UnityVectorExtensions.Epsilon;
+                    }
+                }
+                extra.m_previousDisplacement = displacement;
+                state.PositionCorrection += displacement;
+                extra.confinerDisplacement = displacement.magnitude;
+            }
+        }
+
+        // <summary>
+        /// Calculates Frustum Height for orthographic or perspective camera.
+        /// Ascii illustration of Frustum Height:
+        ///  |----+----+  -\
+        ///  |    |    |    } frustumHeight = cameraWindowHeight / 2
+        ///  |---------+  -/
+        ///  |    |    |
+        ///  |---------|
+        /// </summary>
+        private float CalculateFrustumHeight(in CameraState state, in CinemachineVirtualCameraBase vcam)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = Mathf.Abs(state.Lens.OrthographicSize);
+            }
+            else
+            {
+                Quaternion inverseRotation = Quaternion.Inverse(m_BoundingShape2D.transform.rotation);
+                Vector3 planePosition = inverseRotation * m_BoundingShape2D.transform.position;
+                Vector3 cameraPosition = inverseRotation * vcam.transform.position;
+                float distance = Mathf.Abs(planePosition.z - cameraPosition.z);
+                frustumHeight = distance * Mathf.Tan(state.Lens.FieldOfView * 0.5f * Mathf.Deg2Rad);
+            }
+            return frustumHeight;
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine)
```

Comment:
Do you need to handle confiner offset, as you did in https://github.com/Unity-Technologies/upm-package-cinemachine/pull/108